src/miniscript/iter.rs

```diff
@@ -0,0 +1,552 @@
+pub struct Iter<'a, Pk: 'a + MiniscriptKey, Ctx: 'a + ScriptContext> {
+    next: Option<&'a Miniscript<Pk, Ctx>>,
+    path: Vec<(&'a Miniscript<Pk, Ctx>, usize)>,
+}
+
+impl<'a, Pk: MiniscriptKey, Ctx: ScriptContext> Iter<'a, Pk, Ctx> {
+    fn new(miniscript: &'a Miniscript<Pk, Ctx>) -> Self {
+        Iter {
+            next: Some(miniscript),
+            path: vec![],
+        }
+    }
+}
+
+impl<'a, Pk: MiniscriptKey, Ctx: ScriptContext> Iterator for Iter<'a, Pk, Ctx> {
+    type Item = &'a Miniscript<Pk, Ctx>;
+
+    /// First, the function returns `self`, then the first child of the self (if any),
+    /// then proceeds to the child of the child — down to a leaf of the tree in its first branch.
+    /// When the leaf is reached, it goes in the reverse direction on the same branch until it
+    /// founds a first branching node that had more than a single branch and returns it, traversing
+    /// it with the same algorithm again.
+    ///
+    /// For example, for the given AST
+    /// ```text
+    /// A --+--> B -----> C --+--> D -----> E
+    ///     |                 |
+    ///     |                 +--> F
+    ///     |                 |
+    ///     |                 +--> G --+--> H
+    ///     |                          |
+    ///     |                          +--> I -----> J
+    ///     +--> K
+    /// ```
+    /// `Iter::next()` will iterate over the nodes in the following order:
+    /// `A > B > C > D > E > F > G > I > J > K`
+    ///
+    /// To enumerate the branches iterator uses [Miniscript::branches] function.
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut curr = self.next;
+        if let None = curr {
+            while let Some((node, child)) = self.path.pop() {
+                curr = node.branches().get(child).map(|x| *x);
+                if curr.is_some() {
+                    self.path.push((node, child + 1));
+                    break;
+                }
+            }
+        }
+        if let Some(node) = curr {
+            self.next = node.branches().first().map(|x| *x);
+            self.path.push((node, 1));
+        }
+        curr
+    }
+}
+
+/// Iterator for traversing all [MiniscriptKey]'s in AST starting from some specific node which
+/// constructs the iterator via [Miniscript::iter_pk] method.
+pub struct PkIter<'a, Pk: 'a + MiniscriptKey, Ctx: 'a + ScriptContext> {
+    node_iter: Iter<'a, Pk, Ctx>,
+    keys_buff: VecDeque<Pk>,
+}
+
+impl<'a, Pk: MiniscriptKey, Ctx: ScriptContext> PkIter<'a, Pk, Ctx> {
+    fn new(miniscript: &'a Miniscript<Pk, Ctx>) -> Self {
+        PkIter {
+            node_iter: Iter::new(miniscript),
+            keys_buff: VecDeque::new(),
+        }
+    }
+}
+
+impl<'a, Pk: MiniscriptKey, Ctx: ScriptContext> Iterator for PkIter<'a, Pk, Ctx> {
+    type Item = Pk;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.keys_buff.is_empty() {
+            self.keys_buff = VecDeque::from(loop {
```

Comment:
It seems like all this "compute a `Vec`, convert to `VecDeque`, repeatedly pop" logic could also be replaced with a counter and a `get_nth_pkh` method, eliminating every allocation.

Reply:
Refactored as you've suggested